site/src/contexts/useProxyLatency.ts

```diff
@@ -0,0 +1,152 @@
+import { Region, RegionsResponse } from "api/typesGenerated"
+import { useEffect, useReducer } from "react"
+import PerformanceObserver from "@fastly/performance-observer-polyfill"
+import axios from "axios"
+import { generateRandomString } from "utils/random"
+
+export interface ProxyLatencyReport {
+  // accurate identifies if the latency was calculated using the
+  // PerformanceResourceTiming API. If this is false, then the
+  // latency is calculated using the total duration of the request
+  // and will be off by a good margin.
+  accurate: boolean
+  latencyMS: number
+  // at is when the latency was recorded.
+  at: Date
+}
+
+interface ProxyLatencyAction {
+  proxyID: string
+  report: ProxyLatencyReport
+}
+
+const proxyLatenciesReducer = (
```

Comment:
Why did you decide to use a reducer instead of `useState`? Just curious.

Reply:
The state is a `Record<string, ProxyLatencyReport>` for each proxy in our list. Each proxy record is inserted asynchronously that is triggered from a http request.

This async addition to the state made me concerned about any possible bugs from doing this record insertion manually. It feels like a better design pattern for each proxy callback to just `dispatch` their **own latency report.** 

A `useState` would be like:
```typescript
const [state, setState] = useState<Record<string, ProxyLatencyReport>>({})

const observer = new PerformanceObserver((list) => {
  // If we get entries via this callback, then dispatch the events to the latency reducer.
  list.getEntries().forEach((entry) => {
    setState({
    ...state,
    proxy.id: { /* report */}
    })
  })
})
```

So `useState` just means I am pushing this state management to the caller, where the reducer allows the caller to only be concerned about their personal addition to the state, and they do not have to worry about any previous state.